---
output:
  pdf_document: default
  html_document: default
---
```
Title: "Assessment task 3.Rmd Report, SLE712"
Source: https://github.com/sharmaaad/Final-Assignment-3-SLE712.git
Author: "Aadi Sharma, Student ID 220167207"   
Output type: html_document
Theme: cosmo
```
## Part 1, Question 1. 
Here, we are using the download.file() command to import a "tsv" file from github into R to solve first 5 questions of part 1 of the assessment. To read into R properly, we have formatted the imported data by setting 'header = TRUE', 'stringsAsFactors = FALSE', and 'row.names = 1'.
```{r,chunk1, echo=TRUE}
# downloading the link
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part1_files/gene_expression.tsv", destfile = "gene_expression.tsv")
# reading the data frame in to RStudio
# recognising the header 
# using the first column as the rownames
# avoiding strings to be recognised as Factors
x <-read.table("gene_expression.tsv", header = TRUE, stringsAsFactors = FALSE, row.names = 1)
``` 
### Checking the data integrity
In this section, we are using head()command to read first six lines of the data frame and using str() to make sure that data is recognised as numeric.
```{r, chunk2, echo=TRUE}
# checking the first six lines of the data frame
head(x)
# checking the structure of the data
str(x)
```
## Question 2.
In this part, we are using subset command to make a new column which will be mean of the other columns.

```{r, chunk3, echo=TRUE}
# x is the dataframe
# new column is named 'mean2' to calculate the mean
x$mean2 <- rowMeans(x)
# checking the first six lines of the data
head(x)
```
## Question 3. 
In order to find 10 genes with the highest mean expression, we command the re-ordering of the column named 'mean2'.

```{r, chunk4, echo=TRUE}
# initially, order the column named 'mean2'
order(x$mean2)
# then, by using square brackets, we re-order the column
x[order(-x$mean2),]
x_sorted <- x[order(-x$mean2),]
# in the end, used head() command to find out 10 genes with the highest mean expression
head(x_sorted,10)
```
## Question 4.
In this section, we are filtering the column named 'mean2' to calculate a mean <10. 

```{r, chunk5, echo=TRUE}
# using filter function
filtered <- subset(x, mean2 > 10)
# filtering the number of rows
nrow(filtered)
```
## Question 5.
In this segment, we use the hist() function to make a histogram plot of the mean values of the column named 'mean2' of the provided data. the png form of histogram plot is saved in the repository on github.
```{r, chunk6, echo=TRUE}
# using the hist() command on the column named 'mean2' of the data 'x'
hist(x$mean2)
# to distribute the data frame properly, dividing it by breaks of 30
hist(x$mean2,breaks=30)
```
## Question 6.
Here, we are using the download.file() command to download another "cvs file" from github into R to solve the last 5 questions of part 1 of the assessment. To read into R properly, with read.csv() command, we have formatted the imported data by setting 'header = TRUE', 'stringsAsFactors = FALSE', and 'row.names = 1'.
```{r, chunk7, echo=TRUE}
# a link of another data frame is downloaded from GitHub
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part1_files/growth_data.csv", destfile = "growth_data.csv")
# in order to import "comma separated" values of the data file correctly, we use read.csv() command
y<-read.csv("growth_data.csv")
# reading the data frame into R
# recognising the header
# avoiding strings to be recognised as Factors
y <-read.csv("growth_data.csv", header = TRUE, stringsAsFactors = FALSE)
```
### Checking the data integrity
In this section, we are using str() command to make sure that data is recognised as numeric and we are also naming the columns by using the subset function. 

```{r, chunk8, echo=TRUE}
# checking the structure of the data
str(y)
# to name the columns we are use the subset()command as the data values are 'characters'
subset(y,Site=="northeast")
colnames(y)
```
## Question 7.
Here, we are using subset() command to find out the mean and standard deviation of tree circumference of two different sites which are 'northeast' and 'southwest' from 2004 to 2019.

```{r, chunk9, echo=TRUE}
# subsetting of the tree circumference at the start and end of the study at 'northeast site'
ne<-subset(y, Site == "northeast")
# checking the first six lines of the data frame with head()command
head(ne)
# checking the structure of the data with str()command
str(ne)
# calculating the mean of tree circumference of 'northeast site' from 2004 to 2019
mean(ne$Circumf_2004_cm)
mean(ne$Circumf_2019_cm)
# calculating standard deviation of tree circumference of 'northeast site' from 2004 to 2019
sd(ne$Circumf_2004_cm)
sd(ne$Circumf_2019_cm)

# subsetting of the tree circumference at the start and end of the study at 'southwest site'
sw<-subset(y, Site == "southwest")
# checking the first six lines of the data frame with head()command
head(sw)
# checking the structure of the data with str()command
str(sw)
# the mean of tree circumference of 'southwest site' from 2004 to 2019
mean(sw$Circumf_2004_cm)
mean(sw$Circumf_2019_cm)
# the standard deviation of tree circumference of 'southwest site' from 2004 to 2019
sd(sw$Circumf_2004_cm)
sd(sw$Circumf_2019_cm)
```
## Question 8.
In this part of the assessment task, we are creating a box plot of tree circumference of northeast and southwest sites from 2004 to 2019 by using boxplot() command.
```{r, chunk10, echo=TRUE}
# creating a boxplot of tree circumference of northeast and southwest sites from 2004 to 2019
boxplot(ne$Circumf_2004_cm, ne$Circumf_2019_cm, sw$Circumf_2004_cm, sw$Circumf_2019_cm, names = c("ne2004", "ne2019", "sw2004", "sw2019"), ylab="circumference (m)", las=2, main= "Tree growth at two plantation sites")
```
## Question 9.
Here we are subtracting the tree circumferences from year 2019 to year 2004 (over the past 10 yrars)for both northeast and southwest sites.
```{r, chunk11, echo=TRUE}
# calculating the growth difference by using subtraction command 
ne$growth <-ne$Circumf_2019 - ne$Circumf_2004_cm
# using head () commnad
head(ne)
# using head () commnad
sw$growth <-sw$Circumf_2019 - sw$Circumf_2004_cm
head(sw)
```
## Question 10.
In this section, we are applying the the t.test and wilcox.test functions to estimate the p-value.
```{r, chunk12, echo=TRUE}
# using t.test
t.test(sw$growth, ne$growth)
# using wilcox.test
wilcox.test(sw$growth, ne$growth)
```
## Part 2, Question 1. 
In this part, we download the prerequisite libraries and sources which will be used in R to get the code working. we need seqinr pacakge for processing and analysing the sequence data, R.utils for extracting compressed files and rBLAST for BLAST searches.
```{r, lib}
# downloding the libraries
library("seqinr")
library("R.utils")
library("rBLAST")
library("Biostrings")
# provided source
source("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/mutblast_functions.R")
```
### Importing the whole set of E. coli gene DNA sequences
here, we download the provided file, delete the fast file if any exists by unlinking, uncompress the file by gunziping and eventually create a blast database.
```{r, chunk13, echo=TRUE}
# downloading the provided E. coli CDS file
download.file("ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/cds/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz",
destfile = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz")
# deleting the fasta file if already exists
unlink("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa") 
# uncompressing the file
R.utils::gunzip("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz", overwrite= TRUE)
# creating the blast DB to find out number of sequences present in the E.coli set
makeblastdb("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa",dbtype = "nucl","-parse_seqids")
``` 
## Question 2. Download the sample fasta sequences to determine the length (in bp) and the proportion of GC bases of the allocated sequence. 
In this part, after downloading the fasta file, we are reading the assigned sample sequence '33' into R. Then, checking the structure of the file with str() command to determine the length and the proportion of GC bases of the allocated sequence. 
```{r, chunk15, echo=TRUE}
# importing the file with download.file()command
if ( ! file.exists("sample.fa"))
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/sample.fa", destfile = "sample.fa")  
# reading it in to R
allocatedtask <- read.fasta("sample.fa")
str(allocatedtask)
# reading my sequence 
mytask <- allocatedtask[[33]]
mytask
# checking the data structure of the sequence
str(mytask)
# calculating the GC content with seqinr
seqinr::GC(mytask)
# calculating the length of the sequence (in bp) with seqinr
seqinr::getLength(mytask)
# 0.5300092 GC content and the length of the sequence (in bp) is 1083
```
## Question 3.
Here, we have sourced a function with library()command to run a blast search in order to find the best matches. 

```{r, chunk16, echo=TRUE}
# the function is sourced to run a blast search
library("Biostrings")
source("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/mutblast_functions.R")
# the query sequence and the database are two inputs used for the blast search to obtain the desirable results
myblastn_tab
res <- myblastn_tab(myseq = mytask, db = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa")
# checking the first six lines of the blast search
head(res)
# checking the structure of the data to confirm the presence of numerics and integers
str(res)

# the TOPHIT is 'AAC74295' which belongs to 'prfA gene'.
```

## Question 4. 
In this segment, we are using a mutation function with the mutator() command to note the point mutations in the given sample sequence '33'. By doing so, we have created mutated copies of the sequence. After running the mutator() command few times, we have created a pairwise alignment of the mutated versions of the given sample sequence with the help of the library ("Biostrings").
```{r, chunk17, echo=TRUE}
# creating a mutated version with 30 substitutions
mutator(myseq=mytask, 30)
mytask_mut <- mutator(myseq=mytask, 30)
mytask_mut
# created a mutated version with 100 substitutions
mutator(myseq=mytask, 100)
mytask_mut2 <- mutator(mytask_mut, 100)
mytask_mut2
# now created a pairwise alignment of mutated version and original sample sequence '33' to find number of mismatches
mytask_mut2_ <- DNAString(c2s(mytask_mut2))
mytask_ <- DNAString(c2s(mytask))
aln <- Biostrings::pairwiseAlignment(mytask_,mytask_mut2_)
pid(aln)
nmismatch(aln)
```
## Question 5.
Here, to find the number and proportion of sites that need to be altered to prevent the BLAST search from matching the gene of origin, we have to use functions for mutating and BLASTing the sequence. In order to do so, we need to write the blast index first with the help of write.fasta() command and then, run the blast search with N numbers of mismatches by using the command called nrow(res).
```{r, chunk18, echo=TRUE}
# first need to write the blast index
write.fasta(mytask,names="mytask",file.out = "mytask.fa")
makeblastdb(file="mytask.fa",dbtype = "nucl")
# test with 100 mismatches
mytask_mut <- mutator(myseq=mytask,100)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 130 mismatches
mytask_mut <- mutator(myseq=mytask,130)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 150 mismatches
mytask_mut <- mutator(myseq=mytask,150)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 180 mismatches
mytask_mut <- mutator(myseq=mytask,180)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 200 mismatches
mytask_mut <- mutator(myseq=mytask,200)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 210 mismatches
mytask_mut <- mutator(myseq=mytask,210)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 220 mismatches
mytask_mut <- mutator(myseq=mytask,220)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 250 mismatches
mytask_mut <- mutator(myseq=mytask,250)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 300 mismatches
mytask_mut <- mutator(myseq=mytask,300)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 350 mismatches
mytask_mut <- mutator(myseq=mytask,350)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 400 mismatches
mytask_mut <- mutator(myseq=mytask,400)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 450 mismatches
mytask_mut <- mutator(myseq=mytask,450)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 500 mismatches
mytask_mut <- mutator(myseq=mytask,500)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 550 mismatches
mytask_mut <- mutator(myseq=mytask,550)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 600 mismatches
mytask_mut <- mutator(myseq=mytask,600)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 700 mismatches
mytask_mut <- mutator(myseq=mytask,700)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
# test with 1000 mismatches
mytask_mut <- mutator(myseq=mytask,1000)
res <- myblastn_tab(myseq = mytask_mut, db = "mytask.fa")
res
nrow(res)
```


```{r, chunk19, echo=TRUE}
#created new fuction named 'routine'
routine <- function(myseq, nmut) {
  mutseq <- mutator(myseq = mytask, nmut = nmut)
  res <- myblastn_tab(myseq = mutseq, db="Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa")
  if ( is.null(res) ) { myres=0 } else { myres=1 }
  return(myres)
}


routine(myseq,40)
replicate(35, routine(myseq,40) )


routine(myseq,70)
replicate(35, routine(myseq,70) )
routine(myseq,75)
replicate(35, routine(myseq,75) )

routine(myseq,80)
replicate(35, routine(myseq,80) )

routine(myseq,85)
replicate(35, routine(myseq,85) )

routine(myseq,90)
replicate(35, routine(myseq,90) )
routine(myseq,95)
replicate(35, routine(myseq,95) )


routine(myseq,100)
replicate(50, routine(myseq,100) )

routine(myseq,140)
replicate(80, routine(myseq,140) )

routine(myseq,150)
replicate(100, routine(myseq,150) )

routine(myseq,170)
replicate(110, routine(myseq,170) )

routine(myseq,180)
replicate(120, routine(myseq,180) )

routine(myseq,185)
replicate(125, routine(myseq,185) )

routine(myseq,190)
replicate(130, routine(myseq,190) )

routine(myseq,195)
replicate(135, routine(myseq,195) )

routine(myseq,200)
replicate(140, routine(myseq,200) )
routine(myseq,220)
replicate(145, routine(myseq,220) )

routine(myseq,230)
replicate(150, routine(myseq,230) )

routine(myseq,240)
replicate(150, routine(myseq,240) )

routine(myseq,250)
replicate(160, routine(myseq,250) )

routine(myseq,260)
replicate(165, routine(myseq,260)) 

routine(myseq,265)
replicate(170, routine(myseq,265))

routine(myseq,270)
replicate(175, routine(myseq,270) )

routine(myseq,275)
replicate(180, routine(myseq,275) )

routine(myseq,280)
replicate(185, routine(myseq,280) )

routine(myseq,290)
replicate(190, routine(myseq,290) )

routine(myseq,295)
replicate(195, routine(myseq,295) )


routine(myseq,300)
replicate(200, routine(myseq,300) )

routine(myseq,310)
replicate(210, routine(myseq,310) )

routine(myseq,320)
replicate(220, routine(myseq,320) )

routine(myseq,330)
replicate(230, routine(myseq,330) )

routine(myseq,340)
replicate(240, routine(myseq,340) )

routine(myseq,350)
replicate(250, routine(myseq,350) )

routine(myseq,360)
replicate(260, routine(myseq,360) )

routine(myseq,370)
replicate(265, routine(myseq,370) )

routine(myseq,380)
replicate(270, routine(myseq,380) )

routine(myseq,390)
replicate(280, routine(myseq,390) )

routine(myseq,395)
replicate(285, routine(myseq,395) )


routine(myseq,400)
replicate(290, routine(myseq,400) )

routine(myseq,410)
replicate(300, routine(myseq,410) )

routine(myseq,420)
replicate(310, routine(myseq,420) )

routine(myseq,430)
replicate(320, routine(myseq,430) )

routine(myseq,440)
replicate(330, routine(myseq,440) )

routine(myseq,450)
replicate(340, routine(myseq,450) )

routine(myseq,460)
replicate(350, routine(myseq,460) )

routine(myseq,470)
replicate(360, routine(myseq,470) )

routine(myseq,480)
replicate(370, routine(myseq,480) )

routine(myseq,490)
replicate(380, routine(myseq,490) )

routine(myseq,495)
replicate(385, routine(myseq,495) )


routine(myseq,500)
replicate(390, routine(myseq,500) )

routine(myseq,510)
replicate(400, routine(myseq,510) )

routine(myseq,520)
replicate(410, routine(myseq,520) )

routine(myseq,530)
replicate(420, routine(myseq,530) )

routine(myseq,540)
replicate(430, routine(myseq,540) )

routine(myseq,550)
replicate(440, routine(myseq,550) )

routine(myseq,560)
replicate(450, routine(myseq,560) )

routine(myseq,570)
replicate(460, routine(myseq,570) )

routine(myseq,580)
replicate(470, routine(myseq,580) )

routine(myseq,590)
replicate(480, routine(myseq,590) )

routine(myseq,595)
replicate(485, routine(myseq,595) )


routine(myseq,600)
replicate(490, routine(myseq,600) )

routine(myseq,610)
replicate(500, routine(myseq,610) )

routine(myseq,620)
replicate(510, routine(myseq,620) )

routine(myseq,630)
replicate(520, routine(myseq,630) )

routine(myseq,640)
replicate(530, routine(myseq,640) )

routine(myseq,650)
replicate(540, routine(myseq,650) )

routine(myseq,660)
replicate(550, routine(myseq,660) )

routine(myseq,670)
replicate(560, routine(myseq,670) )

routine(myseq,680)
replicate(570, routine(myseq,680) )

routine(myseq,690)
replicate(580, routine(myseq,690) )

routine(myseq,695)
replicate(585, routine(myseq,695) )


routine(myseq,700)
replicate(590, routine(myseq,700) )

routine(myseq,710)
replicate(600, routine(myseq,710) )

routine(myseq,720)
replicate(610, routine(myseq,720) )

routine(myseq,730)
replicate(620, routine(myseq,730) )

routine(myseq,740)
replicate(630, routine(myseq,740) )

routine(myseq,750)
replicate(640, routine(myseq,750) )

routine(myseq,760)
replicate(650, routine(myseq,760) )

routine(myseq,770)
replicate(660, routine(myseq,770) )

routine(myseq,780)
replicate(670, routine(myseq,780) )

routine(myseq,790)
replicate(680, routine(myseq,790) )

routine(myseq,795)
replicate(685, routine(myseq,795) )


routine(myseq,800)
replicate(690, routine(myseq,800) )

routine(myseq,810)
replicate(700, routine(myseq,810) )

routine(myseq,820)
replicate(710, routine(myseq,820) )

routine(myseq,830)
replicate(720, routine(myseq,830) )


routine(myseq,835)
replicate(725, routine(myseq,835) )

routine(myseq,840)
replicate(730, routine(myseq,840) )

routine(myseq,850)
replicate(740, routine(myseq,850) )

routine(myseq,860)
replicate(750, routine(myseq,860) )

routine(myseq,870)
replicate(760, routine(myseq,870) )

routine(myseq,875)
replicate(765, routine(myseq,875) )


routine(myseq,880)
replicate(770, routine(myseq,880) )

routine(myseq,890)
replicate(780, routine(myseq,890) )



routine(myseq,900)
replicate(790, routine(myseq,900) )

routine(myseq,905)
replicate(795, routine(myseq,905) )

routine(myseq,910)
replicate(800, routine(myseq,910) )

routine(myseq,920)
replicate(810, routine(myseq,920) )

routine(myseq,930)
replicate(820, routine(myseq,930) )

routine(myseq,940)
replicate(830, routine(myseq,940) )

routine(myseq,950)
replicate(840, routine(myseq,950) )

routine(myseq,960)
replicate(850, routine(myseq,960) )

routine(myseq,970)
replicate(860, routine(myseq,970) )

routine(myseq,980)
replicate(870, routine(myseq,980) )

routine(myseq,990)
replicate(880, routine(myseq,990) )

routine(myseq,1000)
replicate(890, routine(myseq,1000) )
```
## question 6
Here we created a line plot and table of data which dipict how the increase of proportion of mutated bases reduces the ability of BLAST to match the gene of origin. 

# Based upon the findings of the chart and the table of the data, it can be said that out of 100 BLAST searches, after the 30ith search, the ability for BLAST to match with the gene of origin went to zero as we increased the proportion of the mutated bases. 
```{r, chunk20, echo=TRUE}
# creating the line plot between nmut values as a vector (x) and the results of BLAST search as a vector(y)
x<-c(35, 50, 80, 100, 110, 120, 125, 130, 135, 140, 145, 150, 150, 160, 165, 170, 175, 180, 185, 190, 195,
     200, 210, 220, 230, 240, 250, 260, 265, 270, 280, 285, 290, 300, 310, 320, 330, 340, 350, 360, 370, 380, 385, 
     390, 400, 410, 420, 430, 440, 450, 460, 470, 480, 485, 490,500, 510, 520, 530, 540, 550, 560, 570, 580, 585, 
     590, 600, 610, 620, 630, 640, 650, 660, 670, 680, 685, 690, 700, 710, 720, 725, 730, 740, 750, 760, 765, 770, 
     780, 790, 795, 800, 810, 820, 830, 840, 850, 860,870, 880, 890)
x
y<- c(1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0,
      1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0)

plot(x,y, xlab="Number of sites randomised", ylab="Proportion of sucessfull BLAST",
     main="How an increase in the number of sites randomised influnce the BLAST performance", 
     type="l", pch="20", lwd="2", col="blue")
```

